# MINST 필기체 숫자 인식

In [1]:
import tensorflow as tf

In [2]:
import numpy as np

In [3]:
from tensorflow import keras

In [4]:
# 신경망과 훈련 매개변수
epochs = 200
batch_size = 128
verbose = 1
nb_classes = 10     # 출력 개수 = 숫자의 개수
n_hidden = 128
validation_split = 0.2     # 검증을 위해 남겨둔 훈련 데이터

In [5]:
# MINST 데이터셋 로드
mnist = keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()

In [6]:
x_train.shape

(60000, 28, 28)

In [7]:
# 이를 28x28에서 60000, 784로 바꿔줌
reshaped = 784
x_train = x_train.reshape(60000, reshaped)
x_test = x_test.reshape(10000, reshaped)

In [8]:
x_train.shape

(60000, 784)

In [9]:
x_train = x_train.astype('float32')
x_test = x_test.astype('float32')

In [10]:
# 픽셀값은 0~255인데 이를 255로 나눠줌으로써 0~1 사이로 정규화시킴
x_train /= 255
x_test /= 255
print(x_train.shape[0], 'train samples')
print(x_test.shape[0], 'test samples')

60000 train samples
10000 test samples


In [11]:
# 원핫 인코딩
y_train = tf.keras.utils.to_categorical(y_train, nb_classes)
y_test = tf.keras.utils.to_categorical(y_test, nb_classes)

In [12]:
dropout = 0.3

In [13]:
# 모델 구축
model = tf.keras.models.Sequential()
model.add(keras.layers.Dense(n_hidden,
                            input_shape = (reshaped,),
                            name = 'dense_layer',
                            activation = 'relu'))
model.add(keras.layers.Dropout(dropout))
model.add(keras.layers.Dense(n_hidden,
                            name = 'dense_layer_2',
                            activation = 'relu'))
model.add(keras.layers.Dropout(dropout))
model.add(keras.layers.Dense(nb_classes,
                            name = 'dense_layer_3',
                            activation = 'softmax'))
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_layer (Dense)          (None, 128)               100480    
_________________________________________________________________
dropout (Dropout)            (None, 128)               0         
_________________________________________________________________
dense_layer_2 (Dense)        (None, 128)               16512     
_________________________________________________________________
dropout_1 (Dropout)          (None, 128)               0         
_________________________________________________________________
dense_layer_3 (Dense)        (None, 10)                1290      
Total params: 118,282
Trainable params: 118,282
Non-trainable params: 0
_________________________________________________________________


## 똑같이 따와서 이부분만 달라짐

In [14]:
# 모델 컴파일
model.compile(optimizer = 'Adam',   # 확률적 그래디언트 하강
             loss = 'categorical_crossentropy',
             metrics = ['accuracy'])

In [15]:
# 모델 훈련
model.fit(x_train, y_train, 
          batch_size = batch_size, 
          epochs = epochs, 
         verbose = verbose, 
         validation_split =  validation_split)

Train on 48000 samples, validate on 12000 samples
Epoch 1/200
48000/48000 [==============================] - 1s 22us/sample - loss: 0.5185 - accuracy: 0.8415 - val_loss: 0.2015 - val_accuracy: 0.9415
Epoch 2/200
48000/48000 [==============================] - 1s 14us/sample - loss: 0.2328 - accuracy: 0.9305 - val_loss: 0.1366 - val_accuracy: 0.9607
Epoch 3/200
48000/48000 [==============================] - 1s 13us/sample - loss: 0.1807 - accuracy: 0.9457 - val_loss: 0.1173 - val_accuracy: 0.9647
Epoch 4/200
48000/48000 [==============================] - 1s 13us/sample - loss: 0.1505 - accuracy: 0.9544 - val_loss: 0.1082 - val_accuracy: 0.9680
Epoch 5/200
48000/48000 [==============================] - 1s 13us/sample - loss: 0.1331 - accuracy: 0.9597 - val_loss: 0.0983 - val_accuracy: 0.9707
Epoch 6/200
48000/48000 [==============================] - 1s 13us/sample - loss: 0.1202 - accuracy: 0.9639 - val_loss: 0.0950 - val_accuracy: 0.9712
Epoch 7/200
48000/48000 [=========================

Epoch 55/200
48000/48000 [==============================] - 1s 13us/sample - loss: 0.0321 - accuracy: 0.9897 - val_loss: 0.0962 - val_accuracy: 0.9794
Epoch 56/200
48000/48000 [==============================] - 1s 13us/sample - loss: 0.0306 - accuracy: 0.9901 - val_loss: 0.1025 - val_accuracy: 0.9788
Epoch 57/200
48000/48000 [==============================] - 1s 13us/sample - loss: 0.0306 - accuracy: 0.9898 - val_loss: 0.1023 - val_accuracy: 0.9786
Epoch 58/200
48000/48000 [==============================] - 1s 13us/sample - loss: 0.0305 - accuracy: 0.9896 - val_loss: 0.1021 - val_accuracy: 0.9797
Epoch 59/200
48000/48000 [==============================] - 1s 13us/sample - loss: 0.0296 - accuracy: 0.9902 - val_loss: 0.1011 - val_accuracy: 0.9796
Epoch 60/200
48000/48000 [==============================] - 1s 13us/sample - loss: 0.0296 - accuracy: 0.9904 - val_loss: 0.1104 - val_accuracy: 0.9790
Epoch 61/200
48000/48000 [==============================] - 1s 13us/sample - loss: 0.0269 - ac

48000/48000 [==============================] - 1s 13us/sample - loss: 0.0199 - accuracy: 0.9940 - val_loss: 0.1145 - val_accuracy: 0.9801
Epoch 110/200
48000/48000 [==============================] - 1s 13us/sample - loss: 0.0212 - accuracy: 0.9931 - val_loss: 0.1140 - val_accuracy: 0.9792
Epoch 111/200
48000/48000 [==============================] - 1s 13us/sample - loss: 0.0223 - accuracy: 0.9931 - val_loss: 0.1123 - val_accuracy: 0.9793
Epoch 112/200
48000/48000 [==============================] - 1s 13us/sample - loss: 0.0221 - accuracy: 0.9928 - val_loss: 0.1129 - val_accuracy: 0.9790
Epoch 113/200
48000/48000 [==============================] - 1s 13us/sample - loss: 0.0213 - accuracy: 0.9927 - val_loss: 0.1132 - val_accuracy: 0.9787
Epoch 114/200
48000/48000 [==============================] - 1s 13us/sample - loss: 0.0209 - accuracy: 0.9933 - val_loss: 0.1171 - val_accuracy: 0.9788
Epoch 115/200
48000/48000 [==============================] - 1s 13us/sample - loss: 0.0236 - accuracy:

Epoch 163/200
48000/48000 [==============================] - 1s 13us/sample - loss: 0.0193 - accuracy: 0.9939 - val_loss: 0.1186 - val_accuracy: 0.9787
Epoch 164/200
48000/48000 [==============================] - 1s 13us/sample - loss: 0.0180 - accuracy: 0.9942 - val_loss: 0.1245 - val_accuracy: 0.9801
Epoch 165/200
48000/48000 [==============================] - 1s 13us/sample - loss: 0.0208 - accuracy: 0.9933 - val_loss: 0.1232 - val_accuracy: 0.9799
Epoch 166/200
48000/48000 [==============================] - 1s 13us/sample - loss: 0.0158 - accuracy: 0.9950 - val_loss: 0.1312 - val_accuracy: 0.9795
Epoch 167/200
48000/48000 [==============================] - 1s 13us/sample - loss: 0.0190 - accuracy: 0.9942 - val_loss: 0.1270 - val_accuracy: 0.9782
Epoch 168/200
48000/48000 [==============================] - 1s 13us/sample - loss: 0.0177 - accuracy: 0.9943 - val_loss: 0.1389 - val_accuracy: 0.9793
Epoch 169/200
48000/48000 [==============================] - 1s 13us/sample - loss: 0.01

In [16]:
# 모델 평가
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy : ',test_acc)

10000/10000 [==============================] - 0s 18us/sample - loss: 0.1287 - accuracy: 0.9803
Test accuracy :  0.9803


### 정확도 98%
### Adam이 가장 좋게 나온다. 최적화기를 선택하는 것은 어떤 경험칙이 있는 것이 아니며 
### 문제 - 최적화기의 조합에 따라 서로 다른 성능을 얻을 수 있다.